File: picker/src/main/java/com/hy/picker/adapter/CrystalAdapter.java
package com.hy.picker.adapter;

import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.facebook.drawee.drawable.ScalingUtils;
import com.facebook.drawee.generic.GenericDraweeHierarchy;
import com.facebook.drawee.view.SimpleDraweeView;
import com.hy.picker.PhotoContext;
import com.hy.picker.PictureSelectorActivity;
import com.hy.picker.R;
import com.hy.picker.core.CrystalResult;
import com.hy.picker.core.ExistBean;
import com.hy.picker.core.util.CrystalDownloadUtils;
import com.hy.picker.core.util.FileUtils;
import com.hy.picker.core.util.SizeUtils;
import com.hy.picker.view.CompletedView;

import androidx.annotation.NonNull;

/**
 * Created time : 2018/8/27 15:53.
 *
 * @author HY
 */
public class CrystalAdapter extends BaseRecyclerAdapter<CrystalResult.Crystal, CrystalAdapter.ViewHolder> {

    private final String cate;
    private final Drawable mDefaultDrawable;

    public CrystalAdapter(String cate, Drawable defaultDrawable) {
        this.cate = cate;
        mDefaultDrawable = defaultDrawable;
    }

    @NonNull
    @Override
    protected ViewHolder createViewHolder(View view, int viewType) {
        return new ViewHolder(view);
    }

    @Override
    protected int layout() {
        return R.layout.picker_item_crystal;
    }


    private int imageSize;

    public class ViewHolder extends BaseRecyclerAdapter.BaseViewHolder {
        private final SimpleDraweeView mIvCrystal;
        private final ImageView mIvDownload;
        private final TextView mTvSize;
        private final TextView mTvWait;
        private final CompletedView mProgress;

        ViewHolder(View itemView) {
            super(itemView);
            mIvCrystal = itemView.findViewById(R.id.picker_iv_crystal);
            mTvSize = itemView.findViewById(R.id.picker_tv_size);
            mProgress = itemView.findViewById(R.id.picker_download_progress);
            mTvWait = itemView.findViewById(R.id.picker_tv_wait);
            mIvDownload = itemView.findViewById(R.id.picker_iv_download);
        }

        @Override
        public void bind() {
            final int position = getAdapterPosition();
            CrystalResult.Crystal item = getItem(position);

            if (imageSize == 0) {
                imageSize = (PhotoContext.getScreenWidth() - 4) / 4 - SizeUtils.dp2px(mContext, 20);
            }

            ViewGroup.LayoutParams layoutParams = mIvCrystal.getLayoutParams();
            if (layoutParams.height != imageSize || layoutParams.width != imageSize) {
                layoutParams.width = imageSize;
                layoutParams.height = imageSize;
            }
            GenericDraweeHierarchy hierarchy = mIvCrystal.getHierarchy();
            hierarchy.setFailureImage(mDefaultDrawable, ScalingUtils.ScaleType.CENTER_CROP);
            hierarchy.setPlaceholderImage(mDefaultDrawable, ScalingUtils.ScaleType.CENTER_CROP);

            mIvCrystal.setController(PictureSelectorActivity
                    .getDraweeController(mIvCrystal, Uri.parse(item.getRes()), imageSize, imageSize));

            mTvSize.setText(FileUtils.formatFileSize(item.getLength()));

            final ExistBean exist = FileUtils.isExist(mContext, cate, item);
            if (exist.isExist()) {
                if (mTvWait.getVisibility() == View.VISIBLE)
                    mTvWait.setVisibility(View.GONE);
                if (mProgress.getVisibility() == View.VISIBLE)
                    mProgress.setVisibility(View.GONE);
                if (mIvDownload.getVisibility() != View.VISIBLE)
                    mIvDownload.setVisibility(View.VISIBLE);
                mIvDownload.setImageResource(R.drawable.picker_complete);
                mIvDownload.setEnabled(false);
            } else {
                mIvDownload.setEnabled(true);
                mIvDownload.setImageResource(R.drawable.picker_download);
                mIvDownload.setOnClickListener(v -> download(position, exist));
            }
            itemView.setOnClickListener(v -> {
                if (exist.isExist()) {
                    if (mOnItemClickListener != null) {
                        mOnItemClickListener.onClick(exist);
                    }
                } else {
                    showDownloadDialog(position, exist);
                }
            });
        }

        private void showDownloadDialog(final int position, final ExistBean exist) {
            new MaterialDialog.Builder(mContext)
                    .title(R.string.picker_title_dialog)
                    .content(R.string.picker_sticker_not_download)
                    .positiveText(android.R.string.ok)
                    .negativeText(R.string.picker_cancel)
                    .onPositive((dialog, which) -> download(position, exist))
                    .show();
        }

        private void download(final int position, ExistBean exist) {
            CrystalResult.Crystal item = getItem(position);
            CrystalDownloadUtils.getInstance()
                    .file(exist.getFile())
                    .url(item.getRes())
                    .length(item.getLength())
                    .download(new CrystalDownloadUtils.DownloadListener() {
                        @Override
                        public void onStart() {
                            if (mTvWait.getVisibility() != View.VISIBLE)
                                mTvWait.setVisibility(View.VISIBLE);
                            if (mProgress.getVisibility() == View.VISIBLE)
                                mProgress.setVisibility(View.GONE);
                            if (mIvDownload.getVisibility() == View.VISIBLE)
                                mIvDownload.setVisibility(View.GONE);
                        }

                        @Override
                        public void onProgress(int progress) {
                            if (mTvWait.getVisibility() == View.VISIBLE)
                                mTvWait.setVisibility(View.GONE);
                            if (mProgress.getVisibility() != View.VISIBLE)
                                mProgress.setVisibility(View.VISIBLE);
                            if (mIvDownload.getVisibility() == View.VISIBLE)
                                mIvDownload.setVisibility(View.GONE);
                            mProgress.setProgress(progress);
                        }

                        @Override
                        public void onSuccess() {
                            notifyItemChanged(position);
                        }

                        @Override
                        public void onFailed() {
                            notifyItemChanged(position);
                        }
                    });
        }

    }

    private OnItemClickListener mOnItemClickListener;

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        mOnItemClickListener = onItemClickListener;
    }

    public interface OnItemClickListener {
        void onClick(ExistBean exist);
    }
}

File: picker/src/main/java/com/hy/picker/adapter/CrystalAdapter.kt
package com.hy.picker.adapter

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.view.View
import androidx.appcompat.app.AlertDialog
import com.facebook.drawee.drawable.ScalingUtils
import com.hy.picker.*
import com.hy.picker.core.Crystal
import com.hy.picker.core.ExistBean
import com.hy.picker.core.util.CrystalDownloadUtils
import com.hy.picker.utils.FileUtils
import kotlinx.android.synthetic.main.picker_item_crystal.*

/**
 * Created time : 2018/8/27 15:53.
 *
 * @author HY
 */
class CrystalAdapter(private val cate: String, private val defaultDrawable: Drawable) :
        PickerBaseRecyclerAdapter<Crystal, CrystalAdapter.ViewHolder>() {

    private val imageSize by lazy{
        (PhotoContext.context.screenWidth() - 4) / 4 - PhotoContext.context.dp(20f)
    }


    override fun createViewHolder(view: View, viewType: Int) = ViewHolder(view)

    override fun layout() = R.layout.picker_item_crystal

    inner class ViewHolder constructor(itemView: View) :
            PickerBaseRecyclerAdapter.BaseViewHolder(itemView) {

        override fun bind() {
            val position = adapterPosition
            val item = getItem(position)


            val layoutParams = pickerIvCrystal.layoutParams
            if (layoutParams.height != imageSize || layoutParams.width != imageSize) {
                layoutParams.width = imageSize
                layoutParams.height = imageSize
            }
            val hierarchy = pickerIvCrystal.hierarchy
            hierarchy.setFailureImage(defaultDrawable, ScalingUtils.ScaleType.CENTER_CROP)
            hierarchy.setPlaceholderImage(defaultDrawable, ScalingUtils.ScaleType.CENTER_CROP)

            pickerIvCrystal.controller = PictureSelectorActivity
                    .getDraweeController(pickerIvCrystal, Uri.parse(item.res), imageSize, imageSize)

            pickerTvSize.text = FileUtils.formatFileSize(item.length.toLong())

            val exist = FileUtils.isExist(itemView.context, cate, item)
            if (exist.exist) {
                if (pickerTvWait.visibility == View.VISIBLE)
                    pickerTvWait.visibility = View.GONE
                if (pickerDownloadProgress.visibility == View.VISIBLE)
                    pickerDownloadProgress.visibility = View.GONE
                if (pickerIvDownload.visibility != View.VISIBLE)
                    pickerIvDownload.visibility = View.VISIBLE
                pickerIvDownload.setImageResource(R.drawable.picker_complete)
                pickerIvDownload.isEnabled = false
            } else {
                pickerIvDownload.isEnabled = true
                pickerIvDownload.setImageResource(R.drawable.picker_download)
                pickerIvDownload.setOnClickListener { download(position, exist) }
            }

            itemView.setOnClickListener {
                if (exist.exist) {
                    _listener?.invoke(exist)
                } else {
                    showDownloadDialog(itemView.context, position, exist)
                }
            }
        }

        private fun showDownloadDialog(context: Context, position: Int, exist: ExistBean) {
            AlertDialog.Builder(context)
                    .setTitle(R.string.pickerTitle_dialog)
                    .setMessage(R.string.picker_sticker_not_download)
                    .setPositiveButton(android.R.string.ok) { dialog, _ ->
                        download(position, exist)
                        dialog.cancel()
                    }
                    .setNegativeButton(R.string.picker_cancel) { dialog, _->
                        dialog.cancel()
                    }
                    .show()
//            MaterialDialog(context)
//                    .title(R.string.pickerTitle_dialog)
//                    .message(R.string.picker_sticker_not_download)
//                    .positiveButton(android.R.string.ok, click = {
//                        download(position, exist)
//                    })
//                    .negativeButton(R.string.picker_cancel)
//                    .show()
        }

        private fun download(position: Int, exist: ExistBean) {
            val item = getItem(position)
            CrystalDownloadUtils.instance
                    .file(exist.file)
                    .url(item.res)
                    .length(item.length)
                    .download(object : CrystalDownloadUtils.DownloadListener {
                        override fun onStart() {
                            if (pickerTvWait.visibility != View.VISIBLE)
                                pickerTvWait.visibility = View.VISIBLE
                            if (pickerDownloadProgress.visibility == View.VISIBLE)
                                pickerDownloadProgress.visibility = View.GONE
                            if (pickerIvDownload.visibility == View.VISIBLE)
                                pickerIvDownload.visibility = View.GONE
                        }

                        override fun onProgress(progress: Int) {
                            if (pickerTvWait.visibility == View.VISIBLE)
                                pickerTvWait.visibility = View.GONE
                            if (pickerDownloadProgress.visibility != View.VISIBLE)
                                pickerDownloadProgress.visibility = View.VISIBLE
                            if (pickerIvDownload.visibility == View.VISIBLE)
                                pickerIvDownload.visibility = View.GONE
                            pickerDownloadProgress.progress=progress
                        }

                        override fun onSuccess() {
                            notifyItemChanged(position)
                        }

                        override fun onFailed() {
                            notifyItemChanged(position)
                        }
                    })
        }

    }

    fun setOnItemClickListener(listener: (ExistBean) -> Unit) {
        _listener = listener
    }

    private var _listener: ((ExistBean) -> Unit)? = null
}
